File: poker_advisor_ui/src/utils/cards.ts
const RANKS = "AKQJT98765432";
const SUITS = "shdc";

export type CardCode = `${string}${string}`;

export function normalizeCard(raw: string): CardCode {
  const token = raw.trim();
  if (token.length !== 2) {
    throw new Error(`Invalid card: ${raw}`);
  }
  const rank = token[0].toUpperCase();
  const suit = token[1].toLowerCase();
  if (!RANKS.includes(rank) || !SUITS.includes(suit)) {
    throw new Error(`Invalid card: ${raw}`);
  }
  return `${rank}${suit}`;
}

export function parseCardList(input: string, expectedCount?: number): CardCode[] {
  const cards = input
    .split(/\s+/)
    .filter(Boolean)
    .map((t) => normalizeCard(t));
  if (expectedCount !== undefined && cards.length !== expectedCount) {
    throw new Error(`Expected ${expectedCount} cards, got ${cards.length}`);
  }
  if (hasDuplicates(cards)) {
    throw new Error("Duplicate cards are not allowed");
  }
  return cards;
}

export function hasDuplicates(cards: Array<CardCode | null | undefined>): boolean {
  const compact = cards.filter((c): c is CardCode => Boolean(c));
  return new Set(compact).size !== compact.length;
}

export function toCardLabel(card: CardCode | null | undefined): string {
  if (!card) {
    return "--";
  }
  const suitMap: Record<string, string> = {
    s: "S",
    h: "H",
    d: "D",
    c: "C",
  };
  return `${card[0]}${suitMap[card[1]] ?? card[1].toUpperCase()}`;
}

export function cardColor(card: CardCode | null | undefined): string {
  if (!card) {
    return "text-slate-400";
  }
  return card[1] === "h" || card[1] === "d" ? "text-rose-300" : "text-slate-100";
}

File: poker_advisor/cards.py
from __future__ import annotations

from typing import Iterable

RANKS = "23456789TJQKA"
SUITS = "cdhs"


class CardParseError(ValueError):
    """Raised when card text cannot be parsed."""


def normalize_card(card: str) -> str:
    token = card.strip()
    if len(token) != 2:
        raise CardParseError(f"Invalid card token: {card!r}")
    rank = token[0].upper()
    suit = token[1].lower()
    if rank not in RANKS or suit not in SUITS:
        raise CardParseError(f"Invalid card token: {card!r}")
    return f"{rank}{suit}"


def parse_cards(text: str, expected_count: int | None = None) -> list[str]:
    tokens = [normalize_card(t) for t in text.split() if t.strip()]
    if expected_count is not None and len(tokens) != expected_count:
        raise CardParseError(f"Expected {expected_count} cards, got {len(tokens)}")
    if len(set(tokens)) != len(tokens):
        raise CardParseError("Duplicate cards provided")
    return tokens


def ensure_unique_cards(cards: Iterable[str]) -> None:
    normalized = [normalize_card(c) for c in cards]
    if len(set(normalized)) != len(normalized):
        raise CardParseError("Duplicate cards found across inputs")


def card_int_to_str(card_value: int, card_cls: object) -> str:
    # treys/deuces both expose Card.int_to_str returning lowercase rank chars.
    raw = card_cls.int_to_str(card_value)
    return raw[0].upper() + raw[1].lower()
